### DISCUSSION TITLE
Using only the to-markdown components of remark-gfm/mdast-util-gfm

### DISCUSSION THREAD

**Discussion by Mr0grog** [remarkjs/remark#1200] [Q&A]
This might be a pretty narrow use case, but I was curious about support for using using *just* the *to* markdown components of remark-gfm or mdast-util-gfm (since by default it sets up both the from and to components). Mainly what I’m concerned about here is byte size in my JS bundle — including just the `gfmToMarkdown` component of `mdast-util-gfm` saves 23 KB in my production JS bundle (about 20% of the overall bundle size at the moment).

Due to the way things are set up right now, the best way I can figure to do this makes me kind of unhappy:

```js
const processor = unified()
  .use(function () {
    // This feels like reaching down into weird implementation details...
    const data = this.data();
    data.toMarkdownExtensions ||= [];
    data.toMarkdownExtensions.push(gfmToMarkdown());
  })
  .use(remarkStringify, {
    // I naively tried setting `extensions: [gfmToMarkdown()]` here
    // before realizing that option gets overridden.
  });
```

First off, I’d love to know if I’m missing something obvious and there’s a more correct way to do this. Some of the constraints I was hoping to adhere to are:
- I’d like to use `mdast-util-gfm` instead of the individual feature extensions if possible, so I don’t have to keep up with new packages being added or removed for GFM.
- It seems good to be using `remark-stringify` instead of `mdast-util-to-markdown` since that’s the recommended approach, and  it does some work to make integration with other plugins just work.

Otherwise, am I worrying to much about these constraints? Should I just use `mdast-util-to-markdown` so I can register extensions directly in its options, and not worry about what I might be missing from `remark-stringify`?

Or would it be reasonable to change one of the involved packages to make this kind of usage easier? (Maybe `remark-stringify` could merge an `extensions` array you pass in directly with the stuff it grabs from `toMarkdownExtensions`? Or maybe `remark-gfm` could have a function to register only the to or only the from Markdown extensions, enabling tree shaking when bundling?

(This was originally posted in the syntax-tree discussion board before I realized I’d posted in the wrong place: https://github.com/orgs/syntax-tree/discussions/110)

**Comment by ChristianMurphy** [comment]
Welcome @Mr0grog! 👋 
Going from AST to compiled document is a more unusual use case, but is supported.
It is one of the reasons `unified` and `remark` are created in such a modular way to allow for different and more custom uses.

> I’d like to use mdast-util-gfm instead of the individual feature extensions if possible, so I don’t have to keep up with new packages being added or removed for GFM.

You can, and you appear to be already doing so above?
Your build tool likely does (or should) include tree shaking, import only the `gfmToMarkdown` function, and the unused functions/code should automatically be excluded.

> I can figure to do this makes me kind of unhappy

What about it makes you unhappy?
`toMarkdownExtensions` is a documented part of the API, it is less commonly seen/used, but is official.

> It seems good to be using `remark-stringify` instead of `mdast-util-to-markdown` since that’s the recommended approach, and it does some work to make integration with other plugins just work.
> Otherwise, am I worrying to much about these constraints? Should I just use mdast-util-to-markdown so I can register extensions directly in its options, and not worry about what I might be missing from remark-stringify?

Up to you, using the `mdast` utilities directly _might_ squeeze a few extra bytes out of the bundle size.
But I personally find the `unified` API easier to work with.

> Or would it be reasonable to change one of the involved packages to make this kind of usage easier? 

What do you have in mind?

> Maybe `remark-stringify` could merge an extensions array you pass in directly with the stuff it grabs from `toMarkdownExtensions`?

I'm not sure I follow, what would be the benefit to this?

> Or maybe remark-gfm could have a function to register only the to or only the from Markdown extensions, enabling tree shaking when bundling?

That already exists, `mdast-util-gfm`

**Reply by Mr0grog** [reply]
> You can, and you appear to be already doing so above? …Your build tool like does (or should) include tree shaking

Yeah, I am using it, and am doing so in order to get tree-shaking. That’s all good. But I feel like the way I’m using it is very hacky. My concern is that there doesn’t seem to be a less-hacky-feeling way of doing it that lets tree-shaking work.

> What about it makes you unhappy?
> `toMarkdownExtensions` is a documented part of the API, it is less commonly seen/used, but is official.

Oh! I was unable to find it documented anywhere (only figured it out by digging around in the source code for all the packages; it’s certainly not in the readme for remark-stringify or mdast-util-gfm), and it seems more like it’s designed for plugin developers rather than end users of the libraries (more my situation). I get what it’s there to support for plugin devs, but it *seems* like someone just trying to plug together extensions probably shouldn’t need be needing it.

>> Or would it be reasonable to change one of the involved packages to make this kind of usage easier?
>
> What do you have in mind?

A few ideas:

1. I imagine the nicest approach from a user perspective would be if remark-gfm exposed `remarkGfmToMarkdown` and `remarkGfmFromMarkdown` alongside the existing `remarkGfm` function, since the existing one prevents you from tree-shaking out either the from or to stuff. You might use it like:

    ```js
    import { remarkGfmToMarkdown } from 'remark-gfm';

    const processor = unified()
      .use(remarkGfmToMarkdown)
      .use(remarkStringify);
    ```

2. Allow `remark-stringify` to take an `extensions` option. This is what I meant when I said:

    > Maybe `remark-stringify` could merge an extensions array you pass in directly with the stuff it grabs from `toMarkdownExtensions`?

    So you could do:

    ```js
    import { gfmToMarkdown } from 'mdast-util-gfm';

    const processor = unified()
      .use(remarkStringify, {
        // Anything in this list gets tacked onto the end of what other plugins have added to `toMarkdownExtensions`:
        extensions: [gfmToMarkdown()]
      });
    ```

    I don’t think it’s quite as nice as approach (1) above, but it’s still fairly straightforward and matches how extensions and other options to the underlying `mdast-util-to-markdown` package works and is publicly documented.

    Right now this doesn’t work because remark-stringify overrides that option, instead of combining it with whatever is in the state’s `toMarkdownExtensions`: https://github.com/remarkjs/remark/blob/1c6e5bac0fee6636380e2d4e5efa62d879f5a0ed/packages/remark-stringify/lib/index.js#L19-L30

>> Or maybe remark-gfm could have a function to register only the to or only the from Markdown extensions, enabling tree shaking when bundling?
>
> That already exists, `mdast-util-gfm`

Am I misreading something? `mdast-util-gfm` doesn’t seem like it can be used as a unified plugin because it doesn’t *register* any of the extensions. It just makes a list of them, and `remark-gfm` does the registering:

https://github.com/syntax-tree/mdast-util-gfm/blob/6e651ee023a285359f9bd0c1427121e4a0d43cdb/lib/index.js#L57-L67

https://github.com/remarkjs/remark-gfm/blob/2d4c9569158afc66dcf1e51ca7d43d680a39c6e8/index.js#L15-L35

**Reply by wooorm** [reply]
2) is supported ;) and documented AFAIK. but it gets complex, hence this is not pushed. https://github.com/syntax-tree/mdast-util-to-markdown/tree/main#optionsextensions and https://github.com/remarkjs/remark/blob/26dc58a675ac7267c105f0fdb76a82db77f8402a/packages/remark-stringify/lib/index.js#L25 (oh wait maybe the last overwrites it instead of merging, nvm)

**Reply by Mr0grog** [reply]
> 2. is supported ;)

Right, with mdast-util-to-markdown, but *not* remark-stringify. That is why I asked originally:

> Otherwise, am I worrying to much about these constraints? Should I just use mdast-util-to-markdown so I can register extensions directly in its options, and not worry about what I might be missing from remark-stringify?

**Reply by Mr0grog** [reply]
> (oh wait maybe the last overwrites it instead of merging, nvm)

Ha, I replied too quick! 😅

**Reply by ChristianMurphy** [reply]
> I was unable to find it documented anywhere (only figured it out by digging around in the source code for all the packages

Improvements to the documentation are welcome!

> it seems more like it’s designed for plugin developers rather than end users of the libraries 

Generally yes, and I think it will stay that way.
It is mostly used by plugin developers adding syntax extensions, a small group within `unified` and `remark` plugin developers.

> but it seems like someone just trying to plug together extensions probably shouldn’t need be needing it.

I'd be interested to hear what other @remarkjs/maintainers think.
My perspective, needing to write a couple lines of code for what I would view as an unusual use case feels like an acceptable trade off, compared with tripling the API surface of `remark-gfm`.

> I imagine the nicest approach from a user perspective would be if `remark-gfm` exposed `remarkGfmToMarkdown` and `remarkGfmFromMarkdown` alongside the existing `remarkGfm` function

I'm cautious as noted above, it feels like adding API surface which could (rarely) add benefit, but would (definitely) increase bundle size for most adopters.
I'm not entirely against the change, but am not yet convinced either.

> Allow `remark-stringify` to take an `extensions` option. This is what I meant when I said:

I'm more open to this, we'd need to establish precedence of data vs options.
But that feels like a smaller change than the rework of `remark-gfm`.

> Am I misreading something? mdast-util-gfm doesn’t seem like it can be used as a unified plugin because it doesn’t register any of the extensions. It just makes a list of them, and remark-gfm does the registering.

The registering is two or three lines of code, versus the slightly abstracted wrapper `remark-gfm` offers is one.
Which is another part of my caution changing `remark-gfm`.
Why add bundle size to several million installs, to add a one line convenience method for a single install?

**Reply by Mr0grog** [reply]
I agree that option (2) is definitely the narrower, less API-surface-impact fix. It’s probably the better of the two solutions in practical terms.

But:

> Why add bundle size to several million installs, to add a one line convenience method for a single install?

I think we’re probably talking about adding a few lines here (`remarkGfm` could be reformulated as just calling the two other functions), so probably like 0.1 KB — and it would save space for people using only the *from* Markdown use case, too. I could be wrong, I imagine a majority of people using `remark-gfm` are only going one direction (whichever direction that is). I don’t think would be benefitting just me, but then again nobody else has asked about this, so maybe nobody else is all that concerned about the 23 KB. 🤷

**Reply by Mr0grog** [reply]
> Improvements to the documentation are welcome!

Happy to do so if helpful, but it sounds like you and @wooorm are saying this is really a plug-in developer feature. …So it doesn't seem like it _should_ be added the end user/README Docs where I was looking, right? So there's probably not actually anything missing?

**Reply by ChristianMurphy** [reply]
> and it would save space for people using only the from Markdown use case, too. I could be wrong, I imagine a majority of people using `remark-gfm` are only going one direction 

`remark` grew out of markdown validation and markdown to markdown transformation.
HTML, XML, React, etc were added later and also have grown into major use cases in their own right.
Markdown to markdown remains major use case.

> I don’t think would be benefitting just me, but then again nobody else has asked about this, so maybe nobody else is all that concerned about the 23 KB.

Again, I am open to the idea as a whole.
The current line of reasoning doesn't work.

Your concerns have centered around documentation being a lot to read and bundle size.
So adding even more documentation to read and adding more code to the bundle size will fix the problem?
I'm not so sure about that.

> so maybe nobody else is all that concerned about the 23 KB.

There is a small but vocal group of developers that care about bundle size above all else, sure.

`unified` and `remark` generally aim to balance developer experience, performance, and bundle size; in that order.
Arguments for bundle size, at the detriment of the other two, tend to not work out as well.

> So it doesn't seem like it should be added the end user/README Docs where I was looking, right?

I'd interested to hear @wooorm's thought.
I'd be find with a mention that the key on the data option can be used.

> So there's probably not actually anything missing?

Taking a few steps back from the singular focus on `remark-gfm`.
There could be a broader guide or recipe to trimming/reducing bundle size on the unified site https://unifiedjs.com/learn/
Covering basics around:
* Avoiding unneeded imports
* Using a tree shaking bundler
* Using a minifier
* Lazy loading (for browser uses, particularly around `react-markdown` and `mdx`)

Could improve developer experience and bundle sizes, without impacting the API or being detrimental to performance.

**Reply by Mr0grog** [reply]
> Your concerns have centered around documentation being a lot to read and bundle size. So adding even more documentation to read and adding more code to the bundle size will fix the problem? I'm not so sure about that.

Hold up a sec here! I *never* suggested the documentation was a lot or too much too read! I spent a lot of time reading docs before posting this, and I have no complaints about the docs. *You suggested I add to the docs,* which I said I’m happy to do (honestly, perfectly happy to!), but based on the overall discussion, it’s not clear that there’s anything *to* document. If the feature in question (`toMarkdownExtensions`) is not for end users, then it makes sense that it’s not in the end-user docs, right?

And, re: bundle size, I made the suggestion because it would *reduce the bundle size* for anybody not using GFM in both directions. I thought one-direction usage would be more common; you set me straight that I was wrong on that point. All good. 👍 

> Arguments for bundle size, at the detriment of the other two, tend to not work out as well.

Just to be clear *I am not asking for that.* I agree, there should be balance.

Like I said in my previous reply, I agree with you that approach (2) is probably a better/more practical approach than (1). If not making any changes at all is the best thing, too, that’s fine. In my original post, I tried to make it clear that telling me I am “worrying to much about these constraints” and should just drop down to the lower level tools is 100% *fine* as an answer. Othewise I’d have just filed an issue instead of discussion.

> Covering basics around… [general bundle size reduction approaches] Could improve developer experience and bundle sizes

Sure, that might be useful. To be clear, though, I’m already doing all of those things. Not knowing them isn’t the barrier here.

Also, that’s a big guide! I am happy to make a smaller docs tweak, I don’t think I really have a huge amount of time to dedicate towards developing something like that. Would be great for someone to write, though.

**Reply by wooorm** [reply]
> I think we’re probably talking about adding a few lines here (remarkGfm could be reformulated as just calling the two other functions), so probably like 0.1 KB — and it would save space for people using only the from Markdown use case, too. I could be wrong, I imagine a majority of people using remark-gfm are only going one direction (whichever direction that is). I don’t think would be benefitting just me, but then again nobody else has asked about this, so maybe nobody else is all that concerned about the 23 KB. 🤷
> — [Rob](https://github.com/orgs/remarkjs/discussions/1200#discussioncomment-6662263)

> I tried to make it clear that telling me I am “worrying to much about these constraints” and should just drop down to the lower level tools is 100% fine as an answer. Othewise I’d have just filed an issue instead of discussion.
> - [Rob](https://github.com/orgs/remarkjs/discussions/1200#discussioncomment-6662615)

If you care about shaving off these bytes, for your particular use case (which is super valid): my recommendation would be to use `mdast-util-to-markdown` with `mdast-util-gfm` directly.
The layer below is great.
The layer above is also great, but supposed to abstract things away.

Plugins exist for two things:
a) to abstract common problems away
b) to let users abstract their problems away

For b) particularly, you can solve it like this:

```js
/**
 * @typedef {import('mdast').Root} Root
 * @typedef {import('mdast-util-gfm').Options} Options
 */
import {gfmToMarkdown} from 'mdast-util-gfm'

/** @type {import('unified').Plugin<[(Options | null | undefined)?], Root>} */
export default function myRemarkGfmToMarkdown(options) {
  const data = this.data()
  const list = /** @type {Array<unknown>} */ (data.toMarkdownExtensions ||= [])
  list.push(gfmToMarkdown(options))
}
```

---

For me, part of the trade off that weighs heavy, is that we often hear that users already find it difficult to figure out how everything works together at the unified/remark/etc layer.
Adding more exports, more docs, will make that worse.
Whereas the alternative, the extra bytes by default or moving a layer lower, seem acceptable to me?

**Reply by Mr0grog** [reply]
Got it.

> For b) particularly, you can solve it like this

This totally works. 👍  I was uncomfortable with this originally (it’s a slightly nicer version of the approach I first posted) because the `toMarkdownExtensions` API seemed like a private API for first-party plugins, but I understand now is that it’s a public, maintained, supported API meant for all plugins in the ecosystem, and I shouldn’t really worry about using it. Great.

Returning briefly to the question of adding support for the `extensions` option to `remark-stringify`, would that still be a good change to make in the long term? It seems like that was left as an open question. That is, support for:

```js
unified()
  .use(somePluginThatAdds_toMarkdownExtensions)
  .use(remarkStringify, { extensions: [ some, other, extensions ] })
```

I feel like it might help resolve one rough mismatch between higher-level plugins and the lower-level parts by letting people use plugins *and* additional extensions together (unlike now, where they must write their own plugin to inject the extensions into `data.toMarkdownExtensions`). I might be missing something there, though. I won’t push on this topic anymore.

> Adding more exports, more docs, will make that worse. Whereas the alternative, the extra bytes by default or moving a layer lower, seem acceptable to me?

I definitely hear this. It makes sense that adding more exports to `remark-gfm` adds complexity in way that wouldn’t be a good tradeoff. I think we got here partly through a misconception on my part about what most common usage looks like, and therefore what makes for a good tradeoff. (Maybe also a misunderstanding of the goals of `remark-gfm` as a package — I understood it as making all the GFM tooling *available through unified,* but really it’s about *abstracting* all the GFM tooling in a single drop-in.)

---

I feel like this thread got kind of argumentative. If the way I phrased anything caused that, I’m sorry; it wasn’t my intent.

I think my response about “I think we’re probably talking about adding a few lines here…” was poorly worded. I was reacting mainly to @ChristianMurphy’s statement that it was “a one line convenience method for a single install,” since I thought lots of people would probably be only converting one direction or the other and would benefit overall. I didn’t make it clear enough that I was responding to that point and was *not* trying to say I thought that made it the best solution. I said that up front, but then spent the balance of my words on something else, which probably made that first line seem disingenuous. Again, sorry for creating a mess there.

**Reply by wooorm** [reply]
> API seemed like a private API for first-party plugins.

It’s fine to use it.
But it’s only meant for folks that are making plugins that integrate with remark-{parse,stringify}/micromark/mdast-util-{to,from}-markdown.
In this round of updates, I’m adding improved support for `Data` sets on nodes, so that it’s typed what is and isn’t there.
I’ll try to set that up for unified’s `Data` too.

> It seems like that was left as an open question. That is, support for:

I’m open to it, but we’d need to figure out precedence: which is prefered over what? `data.settings.fromMarkdownExtensions` first, then `data.fromMarkdownExtensions`, then `options.fromMarkdownExtensions`?

In this update, I’d also want to add symbols to extensions likely.
So that they are deduplicated.
E.g., `data.settings.micromarkExtensions` could include the result of calling `gfmFromMarkdown()`.
But `data.fromMarkdownExtensions` and `options.micromarkExtensions` could too

> I feel like this thread got kind of argumentative.

Discussions about what to add to open source, and how, often do get like that in my experience.
We’re weighing different arguments after all, and figuring out what trade offs to go with.

> If the way I phrased anything caused that, I’m sorry; it wasn’t my intent.

Don’t worry about it!

**Reply by Mr0grog** [reply]
> But it’s only meant for folks that are making plugins that integrate with remark-{parse,stringify}/micromark/mdast-util-{to,from}-markdown.

Side question: it always seemed to me like the general intent was that `remark-*` packages were meant for use with Unified and `mdast-*` packages were not (so `remark-*` APIs conform to the Unified plugin API and might put stuff in or read stuff from `data` and so on, but `mdast-*` would not) — so this usually won’t be a concern of `mdast-*` packages. Is that correct, or am I overgeneralizing from a few examples?

> I’m open to it, but we’d need to figure out precedence: which is prefered over what? `data.settings.fromMarkdownExtensions` first, then `data.fromMarkdownExtensions`, then `options.fromMarkdownExtensions`?

FWIW, my intuition would be that `options.fromMarkdownExtensions` are preferred/added to the end of the list,  so you could override something from a plugin (it looks like the last one always wins: [in maps, only the last is used](https://github.com/syntax-tree/mdast-util-to-markdown/blob/fd6a508cc619b862f75b762dcf876c6b8315d330/lib/configure.js#L44-L52), [`join` is evaluated in reverse and takes the first result](https://github.com/syntax-tree/mdast-util-to-markdown/blob/fd6a508cc619b862f75b762dcf876c6b8315d330/lib/util/container-flow.js#L66-L86), and [`unsafe` looks more complicated, but seems to include all extensions in an order-independent way](https://github.com/syntax-tree/mdast-util-to-markdown/blob/fd6a508cc619b862f75b762dcf876c6b8315d330/lib/util/safe.js#L43-L72)).

I’m sure regular contributors have more informed ideas about what makes sense, though.

**Reply by wooorm** [reply]
> so this usually won’t be a concern of mdast-* packages. Is that correct, or am I overgeneralizing from a few examples?

You are mostly correct. I meant that people who integrate with remark-{parse,stringify} need to know about how micromark and mdast-util-{to,from}-markdown work

> my intuition would be that options.fromMarkdownExtensions are preferred/added to the end of the

Yup, same!
